### ISSUE TITLE
Triggered a suspension while a side effect is still executing

### ISSUE THREAD

**Issue by slinkydeveloper** [restatedev/sdk-typescript#122]
When working on https://github.com/restatedev/e2e/pull/171 I found out a case where a suspension is triggered while a side effect is running. The runtime is configured in always suspending mode.

Consider the following code:

```typescript
const { id, promise } = ctx.awakeable<number[]>();

await ctx.sideEffect(async () => {
   await NumberSortHttpServerUtils.sendSortNumbersRequest(id, numbers);
});

const sortedNumbers: number[] = await promise;
```

You can see by the runtime log that after creating the awakeable the suspension is triggered:

```
2023-07-21T09:13:45.112521Z DEBUG restate_worker::partition::effects
  Effect: Write journal entry Awakeable to storage
    restate.journal.index: 1
  in restate_worker::partition::effects::state_machine_effects
    rpc.service: rnlg.RandomNumberListGenerator
    restate.invocation.sid: rnlg.RandomNumberListGenerator-AYl3uf19eU6qe7r8E6/MFQ==-018977b9fd7d7cae961846ab79f7ef76
[...]
2023-07-21T09:13:45.120668Z TRACE restate_invoker::service::invocation_task
  Received message
    restate.protocol.message_header: MessageHeader { ty: Suspension, length: 3, partial_state_flag: None, protocol_version: None, completed_flag: None, requires_ack_flag: None }
    restate.protocol.message: Suspension(SuspensionMessage { entry_indexes: [1] })
  in restate_invoker::service::invocation_task::invoker_invocation_task
    rpc.system: "restate"
    rpc.service: rnlg.RandomNumberListGenerator
    restate.invocation.sid: rnlg.RandomNumberListGenerator-AYl3uf19eU6qe7r8E6/MFQ==-018977b9fd7d7cae961846ab79f7ef76
2023-07-21T09:13:45.120724Z TRACE restate_invoker::service
  Suspending invocation
  in restate_invoker::service::handle_invocation_task_suspended
    rpc.service: rnlg.RandomNumberListGenerator
    restate.invocation.sid: rnlg.RandomNumberListGenerator-AYl3uf19eU6qe7r8E6/MFQ==-018977b9fd7d7cae961846ab79f7ef76
    restate.invoker.partition_leader_epoch: (7, 1)
2023-07-21T09:13:45.120897Z DEBUG restate_worker::partition::effects
  Effect: Suspend service waiting on entries {1}
    rpc.method: GenerateNumbers
    restate.journal.length: 2
  in restate_worker::partition::effects::state_machine_effects
    rpc.service: rnlg.RandomNumberListGenerator
    restate.invocation.sid: rnlg.RandomNumberListGenerator-AYl3uf19eU6qe7r8E6/MFQ==-018977b9fd7d7cae961846ab79f7ef76
```

But I also observe in the "numbers http server" that the side effect executes twice:

```
2751 [pool-2-thread-1] INFO  dev.restate.e2e.testing.externalhttpserver.Main [] - Got a new request with headers
[...]
3136 [pool-2-thread-1] INFO  dev.restate.e2e.testing.externalhttpserver.Main [] - Got a new request with headers
```

In the case of this test, we then hit this runtime bug https://github.com/restatedev/restate/issues/636 and the test breaks. But even by fixing the runtime bug, I think this is still an issue that can end up locking an invocation forever.

The pattern:

```typescript
const { id, promise } = ctx.awakeable<T>();
await ctx.sideEffect(async () => {
   // Send id somewhere to trigger external action
});
const result: T = await promise;
```

Is the common pattern to use awakeables. If the invocation suspends **before** triggering the external action, the awakeable might never be completed.

My hypothesis is that in this specific test, this doesn't happen because the "await" point inside the side effect is one -> we send a request and then the control flow jumps to the nodejs scheduler, and then we trigger the suspension. My hypothesis is that if we have two requests in the side effect, and only the second one triggers the external action, the invocation ends up stuck. I also think we should verify whether this bug is valid for lambda or not.
